src/Database/Redis/PubSub.hs

```diff
@@ -212,6 +243,343 @@ pubSub initial callback
+-- to all channels in 'PubSubController' will 'addChannelsAndWait' unblock and return.
+addChannelsAndWait :: MonadIO m => PubSubController
+                                -> [(RedisChannel, MessageCallback)] -- ^ the channels to subscribe to
+                                -> [(RedisPChannel, PMessageCallback)] -- ^ the channels to psubscribe to
+                                -> m UnregisterCallbacksAction
+addChannelsAndWait _ [] [] = return $ return ()
+addChannelsAndWait ctrl newChans newPChans = do
+  unreg <- addChannels ctrl newChans newPChans
+  liftIO $ atomically $ do
+    r <- readTVar (pendingCnt ctrl)
+    when (r > 0) retry
+  return unreg
+
+-- | Remove channels from the 'PubSubController', and if there is an active 'pubSubForever', send the
+-- unsubscribe commands to Redis.  Note that as soon as this function returns, no more callbacks will be
+-- executed even if more messages arrive during the period when we request to unsubscribe from the channel
+-- and Redis actually processes the unsubscribe request.  This function is thread-safe.
+--
+-- If you remove all channels, the connection in 'pubSubForever' to redis will stay open and waiting for
+-- any new channels from a call to 'addChannels'.  If you really want to close the connection,
+-- use 'Control.Concurrent.killThread' or 'Control.Concurrent.Async.cancel' to kill the thread running
+-- 'pubSubForever'.
+removeChannels :: MonadIO m => PubSubController
+                            -> [RedisChannel]
+                            -> [RedisPChannel]
+                            -> m ()
+removeChannels _ [] [] = return ()
+removeChannels ctrl remChans remPChans = liftIO $ atomically $ do
+    cm <- readTVar $ callbacks ctrl
+    pm <- readTVar $ pcallbacks ctrl
+    let remChans' = filter (\n -> HM.member n cm) remChans
+        remPChans' = filter (\n -> HM.member n pm) remPChans
+        ps =        (if null remChans' then mempty else unsubscribe remChans')
+          `mappend` (if null remPChans' then mempty else punsubscribe remPChans')
+    writeTBQueue (sendChanges ctrl) ps
+    writeTVar (callbacks ctrl) (foldl' (flip HM.delete) cm remChans')
+    writeTVar (pcallbacks ctrl) (foldl' (flip HM.delete) pm remPChans')
+    modifyTVar (pendingCnt ctrl) (+ totalPendingChanges ps)
+
+-- | Internal function to unsubscribe only from those channels matching the given handle.
+unsubChannels :: PubSubController -> [RedisChannel] -> [RedisPChannel] -> UnregisterHandle -> IO ()
+unsubChannels ctrl chans pchans h = liftIO $ atomically $ do
+    cm <- readTVar $ callbacks ctrl
+    pm <- readTVar $ pcallbacks ctrl
+
+    -- only worry about channels that exist
+    let remChans = filter (\n -> HM.member n cm) chans
+        remPChans = filter (\n -> HM.member n pm) pchans
+
+    -- helper functions to filter out handlers that match
+    let filterHandle :: Maybe [(UnregisterHandle,a)] -> Maybe [(UnregisterHandle,a)]
```

Comment:
what about naming this funciton `removeHandle`? It appears to do a portion of `removeHandles` and doesn't work completely as `filter` so the current name is a bit confusing IMO